### DISCUSSION TITLE
Mixed ride entrance/exit/station styles

### DISCUSSION THREAD

**Discussion by Broxzier** [OpenRCT2/OpenRCT2#15742] [Feature requests]
### Describe the feature

This requires the New Save Format.

Currently, a ride can only have one style which affects the appearance of all its stations, entrances and exits: `Ride::entrance_style`.

It would be nicer to have the tile elements themselves state which style they have instead of storing one style per ride. That way you can have more interesting designs with for example a canvas entrance, abstract exit, and a space station.

### Feature type

Quality of life

**Comment by ocalhoun6** [comment]
Would this only be available through the tile inspector, or how would the UI work for selecting station types per tile?

**Comment by Broxzier** [comment]
It could be implemented in the ride painting tab, much like how you can choose different colours for the vehicles. Quick mock-up:

![image](https://user-images.githubusercontent.com/9705046/137646472-2b118ed7-9520-4255-84f6-b7e3cf2396bc.png)

There could then be an option "Different style per entrance, exit and station" where one can freely set them all individually.

**Comment by frutiemax** [comment]
I'd suggest to make this a vector/array of station styles, which its length is determined by the number of stations for the particular ride.

**Comment by ocalhoun6** [comment]
Ah, so you wouldn't be able to mix and match station styles within the same station then, right? (Like having half the station space themed and then the second half abstract themed.) That's the part that was really boggling me as to how you'd make a user-friendly UI for it.

**Comment by frutiemax** [comment]
That's a good point, it just seems weird to put everything in the Tile though.

**Comment by James103** [comment]
> Ah, so you wouldn't be able to mix and match station styles within the same station then, right? (Like having half the station space themed and then the second half abstract themed.) That's the part that was really boggling me as to how you'd make a user-friendly UI for it.

Would you still be able to do that with the Tile Inspector though? For example, you could use the Tile Inspector to set the station style of the 1st through 4th tiles of the 3rd station to a different style (such as Space), whereas the rest of the station is Abstract style.

**Comment by Broxzier** [comment]
We can definitely expose it in the tile inspector, but it shouldn't be only accessible to players that have cheats enabled. To give more control, there could be a paint mode, similar to how you can repaint parts of the tracks in different colours.

**Comment by duncanspumpkin** [comment]
Track elements have no space for any more information so as it stands it's not possible.

**Comment by Broxzier** [comment]
This feature request is for NSF. Isn't the new save format flexible enough to add new fields?

**Comment by duncanspumpkin** [comment]
No we have already used up all the space.

**Comment by IntelOrca** [comment]
> No we have already used up all the space.

I think we have one byte left.

**Comment by duncanspumpkin** [comment]
> > No we have already used up all the space.
> 
> I think we have one byte left.

Nope that was used for the ride type. We would need two bytes to handle this properly or one byte to handle it acceptably. But there is already a whole bunch of other features that we want to add to the track element so we will have to find some solution.

**Comment by IntelOrca** [comment]
@duncanspumpkin I thought there was one byte left, can you type what you think the structure is with offsets?

**Comment by duncanspumpkin** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/fe46a2cf09825d1dfe7d064c887289d8e8e3d831/src/openrct2/world/TileElement.h#L351

Track type: 5-6
Sequence: 7
Colour scheme: 8
Brake booster speed: 9
Station index: 10
Flags 2: 11
Rideid: 12-13
RideType: 14-15

**Comment by IntelOrca** [comment]
Does ride type need to be two bytes? We haven't even used half of the available numbers yet, if it were a byte.

**Comment by Broxzier** [comment]
Entrance style can also easily fit in one byte.

I find it a little concerning that data is so limited or not flexible enough to add more. I guess that the data of each element still has to have the same size. In that case, data that gets frequently used should be part of the element, and other data could be stored elsewhere, fetched by some unique ID that the elements then have when needed (much like components in ECS).

**Comment by duncanspumpkin** [comment]
> Entrance style can also easily fit in one byte.
> 
> I find it a little concerning that data is so limited or not flexible enough to add more. I guess that the data of each element still has to have the same size. In that case, data that gets frequently used should be part of the element, and other data could be stored elsewhere, fetched by some unique ID that the elements then have when needed (much like components in ECS).

Yes that is what i think we will need to do. Move anything that is for drawing (i.e. not required for logic and accessed super often) to a separate component much like we have with banners.

And no ride type doesnt atm need to be 2 bytes it was done like that for potential future projects.

**Comment by ocalhoun6** [comment]
I'm just an amateur, so maybe I'm wrong... But wouldn't it be a really good idea in the new save format to leave _lots_ of extra room in things like this, even if we currently have no idea what that extra room might be used for?

The last thing we'd want is to roll out the new save format and then later find ourselves saying "No, we can't do that -- it's a limitation of the save format" all over again. (And then eventually have pressure toward a *third* save format.)

Would having generous amounts of unused space in the save format make the performance horribly inefficient or something? Or are we worried about file sizes? I, for one, would be happy to deal with substantially larger file sizes and noticeably longer load/save times if it meant we have ample room in the save format for new features that haven't even been dreamed of yet.

**Comment by duncanspumpkin** [comment]
It's not an issue to change this post nsf. It was always going to be a stepping stone which allowed incremental improvements. The reason this exists is performance and game memory. File size is less of an issue.

**Comment by J-PIE-314** [comment]
This is similar to #21016.
